File: src/python/timing.py
from __future__ import annotations

from dataclasses import dataclass
from time import perf_counter
from typing import List


@dataclass(frozen=True)
class Event:
    label: str
    milliseconds: int


class Stopwatch:
    def __init__(self) -> None:
        now = perf_counter()
        self._start = now
        self._last = now
        self._events: List[Event] = []

    def tick(self, label: str) -> int:
        now = perf_counter()
        elapsed_ms = int((now - self._last) * 1000)
        self._events.append(Event(label, elapsed_ms))
        self._last = now
        return elapsed_ms

    @property
    def events(self) -> List[Event]:
        return list(self._events)

    @property
    def total_ms(self) -> int:
        return int((self._last - self._start) * 1000)

File: src/rust/src/timing.rs
use std::time::Instant;

#[derive(Debug, Clone)]
pub struct Event {
    pub label: String,
    pub milliseconds: u128,
}

#[derive(Debug, Clone)]
pub struct Stopwatch {
    start: Instant,
    last: Instant,
    events: Vec<Event>,
}

impl Stopwatch {
    pub fn new() -> Self {
        let now = Instant::now();
        Self {
            start: now,
            last: now,
            events: Vec::new(),
        }
    }

    pub fn tick<S: Into<String>>(&mut self, label: S) -> u128 {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last).as_millis();
        self.last = now;
        let event = Event {
            label: label.into(),
            milliseconds: elapsed,
        };
        self.events.push(event.clone());
        elapsed
    }

    pub fn events(&self) -> &[Event] {
        &self.events
    }

    pub fn total_ms(&self) -> u128 {
        Instant::now().duration_since(self.start).as_millis()
    }
}
